include/msgpack/adaptor/boost/msgpack_variant.hpp

```diff
@@ -0,0 +1,391 @@
+    boost::string_ref, // STR
+#endif // (BOOST_VERSION / 100000) >= 1 && ((BOOST_VERSION / 100) % 1000) >= 53
+    std::vector<char>, // BIN
+    msgpack::type::raw_ref, // BIN
+    ext,               // EXT
+    ext_ref,           // EXT
+    std::vector<boost::recursive_variant_>, // ARRAY
+    std::map<boost::recursive_variant_, boost::recursive_variant_>, // MAP
+    std::multimap<boost::recursive_variant_, boost::recursive_variant_>, // MAP
+    detail::ref_tag
+>::type variant_ref;
+
+} // namespace type
+
+
+
+
+namespace adaptor {
+
+#if !defined (MSGPACK_USE_CPP03)
+
+template <>
+struct as<msgpack::type::variant> {
+    msgpack::type::variant operator()(msgpack::object const& o) const {
+        switch(o.type) {
+        case type::NIL:
+            return o.as<msgpack::type::nil>();
+        case type::BOOLEAN:
+            return o.as<bool>();
+        case type::POSITIVE_INTEGER:
+            return o.as<uint64_t>();
+        case type::NEGATIVE_INTEGER:
+            return o.as<int64_t>();
+        case type::FLOAT:
+            return o.as<double>();
+        case type::STR:
+            return o.as<std::string>();
+        case type::BIN:
+            return o.as<std::vector<char> >();
+        case type::EXT:
+            return o.as<msgpack::type::ext>();
+        case type::ARRAY:
+            return o.as<std::vector<msgpack::type::variant> >();
+        case type::MAP:
+            return o.as<std::multimap<msgpack::type::variant, msgpack::type::variant> >();
+        default:
+            break;
+        }
+        return msgpack::type::variant();
+    }
+};
+
+#endif // !defined (MSGPACK_USE_CPP03)
+
+
+template <>
+struct convert<msgpack::type::variant> {
+    msgpack::object const& operator()(
+        msgpack::object const& o,
+        msgpack::type::variant& v) const {
+        switch(o.type) {
+        case type::NIL:
+            v = o.as<msgpack::type::nil>();
+            break;
+        case type::BOOLEAN:
+            v = o.as<bool>();
+            break;
+        case type::POSITIVE_INTEGER:
+            v = o.as<uint64_t>();
+            break;
+        case type::NEGATIVE_INTEGER:
+            v = o.as<int64_t>();
+            break;
+        case type::FLOAT:
+            v = o.as<double>();
+            break;
+        case type::STR:
+            v = o.as<std::string>();
+            break;
+        case type::BIN:
+            v = o.as<std::vector<char> >();
+            break;
+        case type::EXT:
+            v = o.as<msgpack::type::ext>();
+            break;
+        case type::ARRAY:
+            v = o.as<std::vector<msgpack::type::variant> >();
+            break;
+        case type::MAP:
+            v = o.as<std::multimap<msgpack::type::variant, msgpack::type::variant> >();
+            break;
+        default:
+            break;
+        }
+        return o;
+    }
+};
+
+namespace detail {
+
+template <typename Stream>
+struct pack_imp:boost::static_visitor<void> {
```

Comment:
Stylistic opinion: I find it hard to tell the actual type name from the inherited class. Putting spaces around the colon after `pack_imp` would make this a lot more readable, imho.